web/openapi/openapi.yaml

```diff
@@ -585,7 +585,7 @@ paths:
   /serve:
     post:
       summary: Return data from a pipeline
-      description: Returns events from an existing pipeline. The pipeline definition must include a serve operator.
+      description: "Returns events from an existing pipeline. The pipeline definition must include a serve operator. By default, the endpoint performs long polling (`timeout: 2s`) and returns events as soon as they are available (`min_events: 1`)"
```

Comment:
This file is auto-generated. The correct place to do this change is `libtenzir/builtins/operators/serve.cpp` and then run `tenzir-ctl web openapi > web/openapi/openapi.yaml`.

Reply:
Sorry for that! Good thing building Tenzir is fast 😅

Reply:
I just did 
```
docker build --target development -t tenzir/tenzir-dev .
docker run --entrypoint tenzir-ctl tenzir/tenzir-dev web openapi > web/openapi/openapi.yaml
```

but the generated `openapi.yaml` doesn't contain the entries for all `/pipeline/xxx` endpoints 😓

Reply:
```
docker run --entrypoint tenzir-ctl ghcr.io/tenzir/tenzir:topic-serve-api-docs web openapi
```

```yaml
---
openapi: 3.0.0
info:
  title: Tenzir Rest API
  version: "\"0.1\""
  description: "\nThis API can be used to interact with a Tenzir Node in a RESTful manner.\n\nAll API requests must be authenticated with a valid token, which must be\nsupplied in the `X-Tenzir-Token` request header. The token can be generated\non the command-line using the `tenzir-ctl web generate-token` command."
servers:
  - url: https://tenzir.example.com/api/v0
security:
  - TenzirToken:
      []
components:
  schemas:
    Metrics:
      type: object
      description: Information about pipeline data ingress and egress.
      properties:
        total:
          type: object
          description: Information about total pipeline metrics (e.g. total ingress/egress)
          properties:
            ingress:
              type: object
              description: Information about the total ingress. Contains no values when the source operator has not run.
              properties:
                unit:
                  type: string
                  enum:
                    - bytes
                    - events
                  description: The unit of the input data.
                  example: bytes
                internal:
                  type: boolean
                  description: Whether the operator's ingress is internal.
                  example: false
                num_elements:
                  type: integer
                  description: The total amount of elements that entered the pipeline source.
                  example: 109834
                num_batches:
                  type: integer
                  description: The total amount of batches that were generated during data processing in the pipeline source.
                  example: 2
                num_approx_bytes:
                  type: integer
                  description: The total amount of bytes that entered the pipeline source. For events, this value is an approximation.
                  example: 30414
                total_seconds:
                  type: number
                  description: The total duration of the pipeline source operation in seconds.
                  example: 1.233998321
                processing_seconds:
                  type: number
                  description: The total duration of the pipeline source data processing in seconds.
                  example: 1.179999992
            egress:
              type: object
              description: Information about the total egress. Contains no values when the sink operator has not run.
              properties:
                unit:
                  type: string
                  enum:
                    - bytes
                    - events
                  description: The unit of the output data.
                  example: bytes
                internal:
                  type: boolean
                  description: Whether the operator's egress is internal.
                  example: false
                num_elements:
                  type: integer
                  description: The total amount of elements that entered the pipeline sink.
                  example: 30414
                num_batches:
                  type: integer
                  description: The total amount of batches that were generated during data processing in the pipeline sink.
                  example: 1
                num_approx_bytes:
                  type: integer
                  description: The total amount of bytes that entered the pipeline sink. For events, this value is an approximation.
                  example: 30414
                total_seconds:
                  type: number
                  description: The total duration of the pipeline sink operation in seconds.
                  example: 2.945935512
                processing_seconds:
                  type: number
                  description: The total duration of the pipeline sink data processing in seconds.
                  example: 1.452123512
    CreateParameters:
      type: object
      required:
        - definition
      properties:
        definition:
          type: string
          example: export | where foo | publish /bar
          description: The pipeline definition.
        name:
          type: string
          description: The human-readable name of the pipeline.
          default: "[an auto-generated id]"
          example: zeek-monitoring-pipeline
        hidden:
          type: boolean
          description: "A flag specifying whether this pipeline is hidden.\nHidden pipelines are not persisted and will not show up in the /pipeline/list endpoint response.\n"
          default: false
          example: false
        ttl:
          type: string
          description: "A duration string specifying the maximum time for this pipeline to exist. No value means the pipeline is allowed to exist forever.\nThis parameter must be defined if the `hidden` parameter is true.\n"
          default: ~
          example: 5.0m
        autostart:
          $ref: "#/components/schemas/PipelineAutostart"
        autodelete:
          $ref: "#/components/schemas/PipelineAutodelete"
    LaunchParameters:
      allOf:
        - $ref: "#/components/schemas/CreateParameters"
        - type: object
          required:
            - serve_id
          properties:
            serve_id:
              type: string
              description: The identifier for the `server` operator.
              example: 4ada2434-32asfe2s
            serve_buffer_size:
              type: integer
              description: The maximum number of events to keep in the `serve` operator.
              example: 4000
    Diagnostics:
      type: array
      items:
        $ref: "#/components/schemas/Diagnostic"
    Diagnostic:
      type: object
      properties:
        severity:
          type: string
          enum:
            - error
            - warning
            - note
        message:
          type: string
          example: unknown option `--frobnify`
        annotation:
          type: array
          items:
            $ref: "#/components/schemas/Annotation"
        notes:
          type: array
          items:
            $ref: "#/components/schemas/Note"
        rendered:
          type: string
          example: "\x1b[1m\x1b[31merror\x1b[39m: unknown option `--frobnify`\x1b[0m\n"
    Annotation:
      type: object
      properties:
        primary:
          type: boolean
        text:
          type: string
          example: this option does not exist
          description: A potentially empty label.
        source:
          $ref: "#/components/schemas/Location"
    Note:
      type: object
      properties:
        kind:
          type: string
          enum:
            - note
            - usage
            - hint
            - docs
          example: usage
        message:
          type: string
          example: file <path> [-f|--follow] [-m|--mmap] [-t|--timeout <duration>]
    Location:
      type: object
      description: A region in the source code, defined by byte offsets.
      properties:
        begin:
          type: number
          example: 42
        end:
          type: number
          example: 48
    PipelineLabel:
      type: object
      properties:
        text:
          type: string
          description: The pipeline label text.
          example: zeek
        color:
          type: string
          description: The pipeline label color.
          example: 3F1A24
    PipelineLabels:
      type: array
      description: The user-provided labels for this pipeline.
      items:
        $ref: "#/components/schemas/PipelineLabel"
    PipelineAutostart:
      type: object
      description: Flags that specify on which state to restart the pipeline.
      properties:
        created:
          type: boolean
          description: Autostart the pipeline upon creation.
          default: false
          example: true
        completed:
          type: boolean
          description: Autostart the pipeline upon completion.
          default: false
          example: false
        failed:
          type: boolean
          description: Autostart the pipeline upon failure.
          default: false
          example: false
        stopped:
          type: boolean
          description: Autostart the pipeline when it stops before completing.
          default: false
          example: true
    PipelineAutodelete:
      type: object
      description: Flags that specify on which state to delete the pipeline.
      properties:
        completed:
          type: boolean
          description: Autodelete the pipeline upon completion.
          default: false
          example: false
        failed:
          type: boolean
          description: Autodelete the pipeline upon failure.
          default: false
          example: true
        stopped:
          type: boolean
          description: Autodelete the pipeline when it stops before completing.
          default: false
          example: false
    PipelineInfo:
      type: object
      properties:
        id:
          type: string
          description: The pipeline ID.
        name:
          type: string
          description: The human-readable name of the pipeline.
        definition:
          type: string
          description: The pipeline definition.
        state:
          type: string
          enum:
            - created
            - running
            - paused
            - failed
            - stopped
        error:
          type: string
          description: The error that the pipeline may have encountered during running.
        diagnostics:
          $ref: "#/components/schemas/Diagnostics"
        metrics:
          $ref: "#/components/schemas/Metrics"
        labels:
          $ref: "#/components/schemas/PipelineLabels"
        autostart:
          $ref: "#/components/schemas/PipelineAutostart"
        autodelete:
          $ref: "#/components/schemas/PipelineAutodelete"
        ttl_expires_in_ns:
          type: integer
          description: If a TTL exists for this pipeline, the remaining TTL in nanoseconds.
          example: 23400569058
  securitySchemes:
    TenzirToken:
      type: apiKey
      in: header
      name: X-Tenzir-Token
paths:
  /ping:
    post:
      summary: Returns a success response
      description: Returns a success response to indicate that the node is able to respond to requests. The response body includes the current node version.
      responses:
        200:
          description: OK.
          content:
            application/json:
              schema:
                type: object
                properties:
                  version:
                    type: string
                    description: The version of the responding node.
                    example: v2.3.0-rc3-32-g8529a6c43f
              example:
                version: v2.3.0-rc3-32-g8529a6c43f
        401:
          description: Not authenticated.
  /pipeline/create:
    post:
      summary: Create a new pipeline
      description: Creates a new pipeline. If `autostart.created` is `true`, the response is only sent after the pipeline was successfully started. If this fails, then the pipeline is immediately discarded, and the response contains the observed diagnostics. The error field then contains a rendered representation of the diagnostics. Otherwise, the diagnostics field is not present.
      requestBody:
        required: true
        content:
          application/json:
            schema:
              $ref: "#/components/schemas/CreateParameters"
      responses:
        200:
          description: Success.
          content:
            application/json:
              schema:
                type: object
                properties:
                  id:
                    type: string
                    description: The ID of the successfully created pipeline.
        400:
          description: Invalid arguments or invalid pipeline.
          content:
            application/json:
              schema:
                type: object
                required:
                  - error
                properties:
                  error:
                    type: string
                    example: Invalid arguments
                    description: The error message.
                  diagnostics:
                    $ref: "#/components/schemas/Diagnostics"
  /pipeline/delete:
    post:
      summary: Delete an existing pipeline
      description: Deletes an existing pipeline.
      requestBody:
        description: Body for the delete endpoint
        required: true
        content:
          application/json:
            schema:
              type: object
              required:
                - id
              properties:
                id:
                  type: string
                  example: 7
                  description: The id of the pipeline to be deleted.
      responses:
        200:
          description: Success.
          content:
            application/json:
              schema:
                type: object
                example:
                  {}
        400:
          description: Invalid arguments.
          content:
            application/json:
              schema:
                type: object
                required:
                  - error
                properties:
                  error:
                    type: string
                    example: Invalid arguments
                    description: The error message.
  /pipeline/launch:
    post:
      summary: Launch a new pipeline
      description: Depending on whether the pipeline ends with a sink, create the pipeline with a [`serve` sink](https://docs.tenzir.com/operators/sinks/serve) with the provided `/pipeline/create` parameters or notify that this pipeline is deployable for further action.
      requestBody:
        required: true
        content:
          application/json:
            schema:
              $ref: "#/components/schemas/LaunchParameters"
      responses:
        200:
          description: Success.
          content:
            application/json:
              schema:
                type: object
                properties:
                  id:
                    type: string
                    description: (If the pipeline does not end with a predefined sink) The ID of the successfully created pipeline.
                  deployable:
                    type: boolean
                    description: (If the pipeline ends with a predefined sink) Returns this property with the value `true`.
  /pipeline/list:
    post:
      summary: List all existing pipelines
      description: Lists all existing pipelines.
      responses:
        200:
          description: Success.
          content:
            application/json:
              schema:
                type: object
                description: An array of all pipelines.
                properties:
                  pipelines:
                    type: array
                    items:
                      $ref: "#/components/schemas/PipelineInfo"
                example:
                  pipelines:
                    - id: 7
                      name: user-assigned-name
                      definition: export | where foo | publish /bar
                      state: running
                      error: ~
                      diagnostics:
                        []
                      metrics:
                        total:
                          in:
                            {}
                          out:
                            {}
                    - id: 8
                      name: wrong-pipeline
                      definition: export asdf
                      state: failed
                      error: format 'asdf' not found
                      diagnostics:
                        []
                      metrics:
                        total:
                          in:
                            {}
                          out:
                            {}
        400:
          description: Invalid arguments.
          content:
            application/json:
              schema:
                type: object
                required:
                  - error
                properties:
                  error:
                    type: string
                    example: Invalid arguments
                    description: The error message.
  /pipeline/reset-ttl:
    post:
      summary: Reset the TTL of an existing pipeline
      description: "Resets the TTL of an existing pipeline as specified, if one has been specified in the /create endpoint before.\nResetting the TTL means that the TTL-related timeout will start counting from zero seconds again, thus keeping the pipeline alive for longer.\n"
      requestBody:
        description: Body for the reset-ttl endpoint
        required: true
        content:
          application/json:
            schema:
              type: object
              required:
                - ids
              properties:
                ids:
                  type: array
                  items:
                    type: string
                  example:
                    - 7
                    - 1
                    - 3
                  description: The id of pipelines whose TTL should be updated.
      responses:
        200:
          description: Success.
          content:
            application/json:
              schema:
                type: object
                properties:
                  ids:
                    type: array
                    items:
                      type: string
                    example:
                      - 7
                      - 1
                    description: The id of pipelines whose TTL has been successfully updated.
        400:
          description: Invalid arguments.
          content:
            application/json:
              schema:
                type: object
                required:
                  - error
                properties:
                  error:
                    type: string
                    example: Invalid arguments
                    description: The error message.
  /pipeline/update:
    post:
      summary: Update pipeline state
      description: "Update the state of the pipeline.\nAll values that are not explicitly included in the request are left in their old state.\n"
      requestBody:
        description: Body for the update endpoint
        required: true
        content:
          application/json:
            schema:
              type: object
              required:
                - id
              properties:
                id:
                  type: string
                  example: 7
                  description: The id of the pipeline to be updated.
                action:
                  type: string
                  enum:
                    - start
                    - pause
                    - stop
                  description: The action that will change the pipeline's running state.
                  example: start
                name:
                  type: string
                  description: Update the human-readable name of the pipeline to this value.
                  example: zeek-monitoring-pipeline
                labels:
                  $ref: "#/components/schemas/PipelineLabels"
      responses:
        200:
          description: Success.
          content:
            application/json:
              schema:
                type: object
                description: The updated pipeline.
                properties:
                  pipeline:
                    $ref: "#/components/schemas/PipelineInfo"
        400:
          description: Invalid arguments.
          content:
            application/json:
              schema:
                type: object
                required:
                  - error
                properties:
                  error:
                    type: string
                    example: Invalid arguments
                    description: The error message.
  /serve:
    post:
      summary: Return data from a pipeline
      description: Returns events from an existing pipeline. The pipeline definition must include a serve operator.
      requestBody:
        description: Body for the serve endpoint
        required: true
        content:
          application/json:
            schema:
              type: object
              required:
                - serve_id
              properties:
                serve_id:
                  type: string
                  example: query1
                  description: The id that was passed to the serve operator.
                continuation_token:
                  type: string
                  example: 340ce2j
                  description: The continuation token that was returned with the last response. For the initial request this is null.
                max_events:
                  type: integer
                  example: 1024
                  default: 1024
                  description: The maximum number of events returned.
                min_events:
                  type: integer
                  example: 1
                  default: 1
                  description: Wait for this number of events before returning.
                timeout:
                  type: string
                  example: 2.0s
                  default: 2.0s
                  description: The maximum amount of time spent on the request. Hitting the timeout is not an error. The timeout must not be greater than 5 seconds.
      responses:
        200:
          description: Success.
          content:
            application/json:
              schema:
                type: object
                properties:
                  next_continuation_token:
                    type: string
                    description: A token to access the next pipeline data batch, null if the pipeline is completed.
                    example: 340ce2j
                  schemas:
                    type: array
                    items:
                      type: object
                      properties:
                        schema_id:
                          type: string
                          description: The unique schema identifier.
                        definition:
                          type: object
                          description: The schema definition in JSON format.
                    description: The schemas that the served events are based on.
                    example:
                      - schema_id: c631d301e4b18f4
                        definition:
                          record:
                            - timestamp: time
                              schema: string
                              schema_id: string
                              events: uint64
                  data:
                    type: array
                    items:
                      type: object
                      properties:
                        schema_id:
                          type: string
                          description: The unique schema identifier.
                        data:
                          type: object
                          description: The actual served data in JSON format.
                    description: The served events.
                    example:
                      - schema_id: c631d301e4b18f4
                        data:
                          timestamp: 2023-04-26T12:00:00.000000
                          schema: zeek.conn
                          schema_id: ab2371bas235f1
                          events: 50
                      - schema_id: c631d301e4b18f4
                        data:
                          timestamp: 2023-04-26T12:05:00.000000
                          schema: suricata.dns
                          schema_id: cd4771bas235f1
                          events: 50
        400:
          description: Invalid arguments.
          content:
            application/json:
              schema:
                type: object
                required:
                  - error
                properties:
                  error:
                    type: string
                    example: Invalid arguments
                    description: The error message.
  /status:
    post:
      summary: Return current status
      description: Returns the current status of the whole node.
      requestBody:
        description: Body for the status endpoint
        required: false
        content:
          application/json:
            schema:
              type: object
              properties:
                component:
                  type: string
                  description: If specified, return the status for that component only.
                  example: index
                verbosity:
                  type: string
                  enum:
                    - info
                    - detailed
                    - debug
                  default: info
      responses:
        200:
          description: OK.
          content:
            application/json:
              schema:
                type: object
              example:
                catalog:
                  num-partitions: 7092
                  memory-usage: 52781901584
                version:
                  Tenzir: v2.3.0-rc3-32-g8529a6c43f
        401:
          description: Not authenticated.
```

Reply:
weird that it isn't working with my local build... I won't try to modify the docs anytime soon 😅

Reply:
The development target is too early, as you need the full CE image to get all the endpoints.